File: image-recognition-v2/app/processors/captioning.py
import asyncio
from transformers import pipeline
from base import Processor  # Importing Processor from base.py

class Captioning(Processor):
    def __init__(self):
        super().__init__()
        self.caption_pipeline = pipeline("image-to-text", model="Xenova/vit-gpt2-image-captioning")

    async def generate_caption(self, image_path: str):
        loop = asyncio.get_event_loop()
        caption = await loop.run_in_executor(self.executor, self.caption_pipeline, image_path)
        return caption

    async def process_image_url(self, image_url: str):
        try:
            caption = await self.generate_caption(image_url)
            return [c['generated_text'] for c in caption]
        except Exception as e:
            # Handle exceptions or propagate them
            raise

File: image-recognition-v1/src/processors/captioning.ts
import { pipeline, ImageToTextOutput, ImageToTextPipeline } from "@xenova/transformers";
import { Processor } from "./base.js";
import _ from "lodash";

export class Captioning extends Processor {
  pipeline: ImageToTextPipeline;
  constructor(){
    super();
    let percentages: number[] = [];
    pipeline('image-to-text', 'Xenova/vit-gpt2-image-captioning', {
      progress_callback: (progress: {progress: number, loaded: number, total: number}) => {
        // should be every like 10% or something. problem is, it's not an integer so...
        if (!_.isNil(progress.progress) && percentages.indexOf(Math.round(progress.progress)) === -1 && progress.progress < 100) {
          percentages.push(Math.round(progress.progress));
          console.log(`Downloading Xenova/vit-gpt2-image-captioning: ${Math.ceil(progress.progress)}%`);
        }
      }
    })
      .then((pipe) => {
        this.pipeline = pipe;
        this.setupDeferrer.resolve();
      })
  }

  public async processImageUrl(imageUrl: string): Promise<string[]> {
    await this.setupDeferrer;
    const output = await this.pipeline(imageUrl);
    const wrappedOutput = (Array.isArray(output) ? output : [output]) as ImageToTextOutput;
    return wrappedOutput.map((caption) => caption.generated_text);
  }
}

export default new Captioning();